controllers/topology/current_state.go

```diff
@@ -18,12 +18,132 @@ package topology
+		return nil, err
+	}
+	clusterState.machineDeployments = m
+	return &clusterState, nil
+}
+
+// getCurrentInfrastructureClusterState looks for the state of the InfrastructureCluster. If a reference is set but not
+// found, either from an error or the object not being found, an error is thrown.
+func (r *ClusterReconciler) getCurrentInfrastructureClusterState(ctx context.Context, cluster *v1alpha4.Cluster) (*unstructured.Unstructured, error) {
+	infra, err := r.getTemplate(ctx, cluster.Spec.InfrastructureRef)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to read %s %s", cluster.Spec.InfrastructureRef.Kind, cluster.Spec.InfrastructureRef.Name)
+	}
+	return infra, nil
+}
+
+// getCurrentControlPlaneState returns information on the ControlPlane being used by the Cluster. If a reference is not found,
+// an error is thrown. If the ControlPlane requires MachineInfrastructure according to its ClusterClass an error will be
+// thrown if the ControlPlane has no MachineTemplates.
+func (r *ClusterReconciler) getCurrentControlPlaneState(ctx context.Context, cluster *v1alpha4.Cluster, class v1alpha4.ClusterClass) (controlPlaneTopologyState, error) {
+	cp, err := r.getTemplate(ctx, cluster.Spec.ControlPlaneRef)
+	if err != nil {
+		return controlPlaneTopologyState{}, errors.Wrapf(err, "failed to read %s %s", cluster.Spec.ControlPlaneRef.Kind, cluster.Spec.ControlPlaneRef.Name)
+	}
+	// Some ControlPlane providers may not require MachineInfrastructure to run. This check returns early if the field
+	// indicating MachineInfrastructure is required is not found in the ClusterClass of the given Cluster.
+	if class.Spec.ControlPlane.MachineInfrastructure == nil {
+		return controlPlaneTopologyState{
+			cp,
+			nil,
+		}, nil
+	}
+
+	cpInfraRef, err := getNestedRef(cp, "spec", "machineTemplate", "infrastructureRef")
+	if err != nil {
+		return controlPlaneTopologyState{cp, nil}, errors.Wrapf(err, "failed to get InfrastructureMachineTemplate reference for %s, %s", cp.GetKind(), cp.GetName())
+	}
+	infra, err := r.getTemplate(ctx, cpInfraRef)
+	if err != nil {
+		return controlPlaneTopologyState{cp, nil}, errors.Wrapf(err, "failed to get InfrastructureMachineTemplate reference for %s, %s", cp.GetKind(), cp.GetName())
+	}
+	return controlPlaneTopologyState{
+		cp,
+		infra,
+	}, nil
+}
+
+// getCurrentMachineDeploymentState queries for all MachineDeployments and filters them for their linked Cluster and
+// whether they are managed by a ClusterClass using labels. A Cluster may have zero or more MachineDeployments. Zero is
+// expected on first reconcile. If MachineDeployments are found for the Cluster their Infrastructure and Bootstrap references
+// are inspected. Where these are not found the function will throw an error.
+func (r *ClusterReconciler) getCurrentMachineDeploymentState(ctx context.Context, cluster *v1alpha4.Cluster) ([]machineDeploymentTopologyState, error) {
+	mDeploymentState := make([]machineDeploymentTopologyState, 0)
```

Comment:
According to changes in https://github.com/kubernetes-sigs/cluster-api/pull/5070 and in https://github.com/kubernetes-sigs/cluster-api/pull/5072 it seems that it is much more effective if we store desired/current state for machine deployments as `map[string]machineDeploymentTopologyState`, with the key being the value read from the https://github.com/kubernetes-sigs/cluster-api/blob/e4e68c1249648aee68fdd4e044f1b23d0b262ecc/api/v1alpha4/common_types.go#L34 label

NOTE: if a md does not have this label we should error out

@killianmuldoon do you think you can make this change within this PR?
@sbueringer is it ok for you?

Reply:
I think it should be resolved in this PR - I'll take a look at doing so now.

Reply:
Yup @srm09 also proposed it here: https://github.com/kubernetes-sigs/cluster-api/pull/5070#discussion_r686170364

I'm fine with just aligning to the changes from either here or #5070 (or both :))

Reply:
> I think it should be resolved in this PR - I'll take a look at doing so now.

agreed, thanks @killianmuldoon